File: problems/problems_994/solution.ts
import {Queue} from "@datastructures-js/queue";

function orangesRotting(grid: number[][]): number {
    const m: number = grid.length, n: number = grid[0].length;
	const queue: Queue<number[]> = new Queue<number[]>();
	let fresh: number = 0, times: number = 0;
	const directions: number[][] = [[0, 1], [0, -1], [1, 0], [-1, 0]];
	for (let i: number = 0; i < m; i++) {
		for (let j: number = 0; j < n; j++) {
			if (grid[i][j] === 2) {
				queue.enqueue([i, j]);
			} else if (grid[i][j] === 1) {
				fresh++;
			}
		}
	}
	if (fresh === 0) return 0;
	while (!queue.isEmpty()) {
		times++;
		const size: number = queue.size();
		for (let i: number = 0; i < size; i++) {
			const [x, y]: number[] = queue.dequeue();
			for (const [dx, dy] of directions) {
				const nx: number = x + dx, ny: number = y + dy;
				if (nx < 0 || nx >= m || ny < 0 || ny >= n || grid[nx][ny] !== 1) continue;
				grid[nx][ny] = 2;
				fresh--;
				queue.enqueue([nx, ny]);
			}
		}
	}
	return fresh === 0 ? times - 1 : -1;
};

export function Solve(inputJsonElement: string): any {
	const inputValues: string[] = inputJsonElement.split("\n");
	const grid: number[][] = JSON.parse(inputValues[0]);
	return orangesRotting(grid);
}

File: problems/problems_994/solution.go
package problem994

import (
	"encoding/json"
	"log"
	"strings"
)

type pair struct{ x, y int }

var directions = []pair{{-1, 0}, {1, 0}, {0, -1}, {0, 1}} // 四方向

func orangesRotting(grid [][]int) int {
	m, n := len(grid), len(grid[0])
	fresh := 0
	q := []pair{}
	for i, row := range grid {
		for j, x := range row {
			if x == 1 {
				fresh++ // 统计新鲜橘子个数
			} else if x == 2 {
				q = append(q, pair{i, j}) // 一开始就腐烂的橘子
			}
		}
	}

	ans := -1
	for len(q) > 0 {
		ans++ // 经过一分钟
		tmp := q
		q = []pair{}
		for _, p := range tmp { // 已经腐烂的橘子
			for _, d := range directions { // 四方向
				i, j := p.x+d.x, p.y+d.y
				if 0 <= i && i < m && 0 <= j && j < n && grid[i][j] == 1 { // 新鲜橘子
					fresh--
					grid[i][j] = 2 // 变成腐烂橘子
					q = append(q, pair{i, j})
				}
			}
		}
	}

	if fresh > 0 {
		return -1
	}
	return max(ans, 0)
}

func Solve(input string) any {
	values := strings.Split(input, "\n")
	var grid [][]int

	if err := json.Unmarshal([]byte(values[0]), &grid); err != nil {
		log.Fatal(err)
	}

	return orangesRotting(grid)
}
